docs/proposals/20191017-kubeadm-based-control-plane.md

```diff
@@ -357,23 +367,107 @@ spec:
+  - The cluster MUST have no deleting machines. This rule prevents KCP taking actions while the cluster is in a transitional state.
+  - Remediation MUST preserve etcd quorum. This rule ensures that we will not remove a member that would result in etcd
+    losing a majority of members and thus become unable to field new requests.
+
+- When all the conditions for stating remediaiton are satisfied, KCP temporarily suspend any operation in progress
+  in order to perform remediation. 
+- Remediation will be performed by issuing a delete on the unhealthy machine; after deleting the machine, KCP
+  will restore the target number of machines by triggering a scale up (current replicas<desired replicas) and then
+  eventually restart the rollout action.
+  See [why delete and recreate](#why-delete-and-recreate) for an explanation about why KCP should remove the member first and then add its replacement.
+
+###### Why delete and recreate
+
+When replacing a KCP machine the most critical component to be taken into account is etcd, and
+according to the [etcd documentation](https://github.com/etcd-io/etcd/blob/master/Documentation/faq.md#should-i-add-a-member-before-removing-an-unhealthy-member),
+it's important to remove an unhealthy etcd member first and then add its replacement:
+
+_etcd employs distributed consensus based on a quorum model; (n/2)+1 members, a majority, must agree on a proposal before
+it can be committed to the cluster. These proposals include key-value updates and membership changes. This model totally
+avoids any possibility of split brain inconsistency. The downside is permanent quorum loss is catastrophic._
+
+_How this applies to membership: If a 3-member cluster has 1 downed member, it can still make forward progress because
+the quorum is 2 and 2 members are still live. However, adding a new member to a 3-member cluster will increase the quorum
+to 3 because 3 votes are required for a majority of 4 members. Since the quorum increased, this extra member buys nothing
+in terms of fault tolerance; the cluster is still one node failure away from being unrecoverable._
+
+_Additionally, that new member is risky because it may turn out to be misconfigured or incapable of joining the cluster.
+In that case, there's no way to recover quorum because the cluster has two members down and two members up, but needs
+three votes to change membership to undo the botched membership addition. etcd will by default reject member
+add attempts that could take down the cluster in this manner._
+
+_On the other hand, if the downed member is removed from cluster membership first, the number of members becomes 2 and
+the quorum remains at 2. Following that removal by adding a new member will also keep the quorum steady at 2. So, even
+if the new node can't be brought up, it's still possible to remove the new member through quorum on the remaining live members._
```

Comment:
We'll need to make sure to prioritize nodes that have unhealthy etcd members instead of any or the oldest unhealthy machine to achieve this

Reply:
see https://github.com/kubernetes-sigs/cluster-api/pull/3676#discussion_r492975642